fix #45552 - adding add_done_callback(fn) to torch.futures.Future

bdhirsh opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45675 fix #45552 - adding add_done_callback(fn) to torch.futures.Future**

Differential Revision: [D24055353](https://our.internmc.facebook.com/intern/diff/D24055353)

bdhirsh committed:
fix #45552 - adding add_done_callback(fn) to torch.futures.Future

[ghstack-poisoned]

bdhirsh committed:
Update on "fix #45552 - adding add_done_callback(fn) to torch.futures.Future"


Differential Revision: [D24055353](https://our.internmc.facebook.com/intern/diff/D24055353)

[ghstack-poisoned]

mrshenli commented on torch/futures/__init__.py:
```diff
@@ -81,6 +81,39 @@ def then(self, callback):  # type: (Callable[[Future[T]], S]) -> Future[S]
+        Multiple callbacks can be added to the same ``Future``, and all
+        added callbacks will be executed once the future completes.
+        If the future has already completed, the callback will be executed immediately.
+        The callback
```
unfinished doc?

mrshenli commented on torch/futures/__init__.py:
```diff
@@ -81,6 +81,39 @@ def then(self, callback):  # type: (Callable[[Future[T]], S]) -> Future[S]
         """
         return cast(Future[S], super().then(callback))
 
+    def add_done_callback(self, callback):  # type: (Callable[[None], None]) -> None
```
Let's comply with [Python's `add_done_callback` API](https://docs.python.org/3/library/asyncio-future.html#asyncio.Future.add_done_callback), which says

> The callback is called with the Future object as its only argument.

mrshenli commented on test/test_futures.py:
```diff
@@ -92,6 +92,41 @@ def test_chained_then(self):
+        self.assertEqual(fut.wait(), torch.ones(2, 2))
+        self.assertTrue(callback_result)
+
+    def test_add_done_callback_maintains_callback_order(self):
```
Let's add a test to verify that `add_done_callback` and `then` can be used together and runs in order.

bdhirsh replied:
agreed, done

mrshenli reviewed (changes requested):
Thanks for putting this up so promptly.

bdhirsh committed:
Update on "fix #45552 - adding add_done_callback(fn) to torch.futures.Future"


Differential Revision: [D24055353](https://our.internmc.facebook.com/intern/diff/D24055353)

[ghstack-poisoned]

bdhirsh committed:
Update on "fix #45552 - adding add_done_callback(fn) to torch.futures.Future"


Differential Revision: [D24055353](https://our.internmc.facebook.com/intern/diff/D24055353)

[ghstack-poisoned]

pritamdamania87 reviewed (commented):
I'm a bit late on this, but I'm wondering what value this API brings on top of `.then()`? Can't we achieve something similar with `.then()` as well?

My feeling is having a smaller API surface makes it easier in general to manage things longer term.

bdhirsh committed:
Update on "fix #45552 - adding add_done_callback(fn) to torch.futures.Future"


Differential Revision: [D24055353](https://our.internmc.facebook.com/intern/diff/D24055353)

[ghstack-poisoned]

bdhirsh committed:
Update on "fix #45552 - adding add_done_callback(fn) to torch.futures.Future"


Differential Revision: [D24055353](https://our.internmc.facebook.com/intern/diff/D24055353)

[ghstack-poisoned]

bdhirsh committed:
Update on "fix #45552 - adding add_done_callback(fn) to torch.futures.Future"


Differential Revision: [D24055353](https://our.internmc.facebook.com/intern/diff/D24055353)

[ghstack-poisoned]

mrshenli commented:
Hey @pritamdamania87 this is requested by @jjlilley a few months ago in #37908, and this is also closer to the Python Future API. If we are not confident whether this will be useful, I guess we can mark this as a private API for now?

mrshenli commented on torch/testing/_internal/distributed/rpc/jit/rpc_test.py:
```diff
@@ -1038,6 +1038,23 @@ def callback(fut):
+            args=(torch.ones(2),),
+        ).add_done_callback(callback)
+        self.assertFalse(callback_called)
+        self.assertEqual(future.wait(), torch.ones(2) * 2)
```
hmm, future is None, as it is the return value of `add_done_callback`, right?

mrshenli commented on torch/csrc/jit/python/pybind_utils.h:
```diff
@@ -144,6 +144,16 @@ struct VISIBILITY_HIDDEN PythonFutureWrapper
+    fut->addCallback(std::bind(
+        [pyFut(this->getPtr())](std::shared_ptr<PythonFunctionGuard> pf) {
+          pybind11::gil_scoped_acquire ag;
+          pf->func_(pyFut);
```
what if this throws error? can we add a test to verify the behavior?

bdhirsh replied:
Done, thanks for the reminder (had to double check the API). Added logic to swallow and print any exceptions thrown in the callback + a test.

mrshenli commented on torch/testing/_internal/distributed/rpc/jit/rpc_test.py:
```diff
@@ -1038,6 +1038,23 @@ def callback(fut):
+        ).add_done_callback(callback)
+        self.assertFalse(callback_called)
+        self.assertEqual(future.wait(), torch.ones(2) * 2)
+        self.assertEqual(callback_called, torch.ones(2) * 4)
```
If you let `future` be the return value of `rpc_async`, then the completion of `future` no longer guarantees the completion of the callback. You will need additional signals to indicate the callback is done.

bdhirsh replied:
Right, thanks. I added `then` callbacks to those tests that should run after the `add_done_callback` which I now wait on. That should guarantee they run before the test finishes- it makes the test a little more convoluted, but given that we have separated tests devoted just to testing `then` I think that's fine.

rohan-varma replied:
Nit: I think that this approach is fine, although it implicitly assumes an implementation detail, which is that the callbacks are called in the order that they are added (but the only time the order should matter to the user is when they explicitly chain callbacks with `.then(...).then(....)` etc. 

I think it's fine to assume this for now since the implementation is probably unlikely to change, although another way to test this could be to have the done callback set a `threading.Event` and wait on that.

bdhirsh replied:
It looks like we do according to the docs here? https://github.com/pytorch/pytorch/blob/8a1e1004664e4cc8b1adbcef07b9fdae397e1787/torch/futures/__init__.py#L49

At least, it sounds like we make guarantees around the order that callbacks registered from `then` are called. And we have the same guarantee for callbacks registered from `add_done_callback`. I guess the fact that any "interleaving" of these calls will also be called in the same registration order is really an implementation detail, although that will probably be unlikely to change too. Either way thanks for offering the alternative.

bdhirsh committed:
Update on "fix #45552 - adding add_done_callback(fn) to torch.futures.Future"


Differential Revision: [D24055353](https://our.internmc.facebook.com/intern/diff/D24055353)

[ghstack-poisoned]

bdhirsh committed:
Update on "fix #45552 - adding add_done_callback(fn) to torch.futures.Future"


Differential Revision: [D24055353](https://our.internmc.facebook.com/intern/diff/D24055353)

[ghstack-poisoned]

bdhirsh committed:
Update on "fix #45552 - adding add_done_callback(fn) to torch.futures.Future"


Differential Revision: [D24055353](https://our.internmc.facebook.com/intern/diff/D24055353)

[ghstack-poisoned]

mrshenli reviewed (changes requested):
Looks like the new test is flaky:

```
Oct 02 19:35:39   test_add_done_callback (__main__.ProcessGroupJitRpcTestWithSpawn) ... RPC was initialized with the PROCESS_GROUP backend which is deprecated and slated to be removed and superseded by the TENSORPIPE backend. It is recommended to migrate to the TENSORPIPE backend.
Oct 02 19:35:39 RPC was initialized with the PROCESS_GROUP backend which is deprecated and slated to be removed and superseded by the TENSORPIPE backend. It is recommended to migrate to the TENSORPIPE backend.
Oct 02 19:35:39 RPC was initialized with the PROCESS_GROUP backend which is deprecated and slated to be removed and superseded by the TENSORPIPE backend. It is recommended to migrate to the TENSORPIPE backend.
Oct 02 19:35:39 RPC was initialized with the PROCESS_GROUP backend which is deprecated and slated to be removed and superseded by the TENSORPIPE backend. It is recommended to migrate to the TENSORPIPE backend.
Oct 02 19:35:39 ERROR:root:Caught exception: 
Oct 02 19:35:39 Traceback (most recent call last):
Oct 02 19:35:39   File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/testing/_internal/common_distributed.py", line 259, in wrapper
Oct 02 19:35:39     fn()
Oct 02 19:35:39   File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/testing/_internal/dist_utils.py", line 90, in new_test_method
Oct 02 19:35:39     return_value = old_test_method(self, *arg, **kwargs)
Oct 02 19:35:39   File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/testing/_internal/distributed/rpc/jit/rpc_test.py", line 1059, in test_add_done_callback
Oct 02 19:35:39     self.assertEqual(callback_called, None)
Oct 02 19:35:39   File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/testing/_internal/common_utils.py", line 1175, in assertEqual
Oct 02 19:35:39     super().assertEqual(x, y, msg=msg)
Oct 02 19:35:39   File "/Users/distiller/workspace/miniconda3/lib/python3.7/unittest/case.py", line 852, in assertEqual
Oct 02 19:35:39     assertion_func(first, second, msg=msg)
Oct 02 19:35:39   File "/Users/distiller/workspace/miniconda3/lib/python3.7/unittest/case.py", line 845, in _baseAssertEqual
Oct 02 19:35:39     raise self.failureException(msg)
Oct 02 19:35:39 AssertionError: tensor([4., 4.]) != None
Oct 02 19:35:39 exiting process with exit code: 10
Oct 02 19:35:39 ERROR:root:Caught exception: 
Oct 02 19:35:39 Traceback (most recent call last):
Oct 02 19:35:39   File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/testing/_internal/common_distributed.py", line 259, in wrapper
Oct 02 19:35:39     fn()
Oct 02 19:35:39   File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/testing/_internal/dist_utils.py", line 90, in new_test_method
Oct 02 19:35:39     return_value = old_test_method(self, *arg, **kwargs)
Oct 02 19:35:39   File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/testing/_internal/distributed/rpc/jit/rpc_test.py", line 1059, in test_add_done_callback
Oct 02 19:35:39     self.assertEqual(callback_called, None)
Oct 02 19:35:39   File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/testing/_internal/common_utils.py", line 1175, in assertEqual
Oct 02 19:35:39     super().assertEqual(x, y, msg=msg)
Oct 02 19:35:39   File "/Users/distiller/workspace/miniconda3/lib/python3.7/unittest/case.py", line 852, in assertEqual
Oct 02 19:35:39     assertion_func(first, second, msg=msg)
Oct 02 19:35:39   File "/Users/distiller/workspace/miniconda3/lib/python3.7/unittest/case.py", line 845, in _baseAssertEqual
Oct 02 19:35:39     raise self.failureException(msg)
Oct 02 19:35:39 AssertionError: tensor([4., 4.]) != None
Oct 02 19:35:39 exiting process with exit code: 10
Oct 02 19:35:39 Process 1 terminated with exit code 10, terminating remaining processes.
Oct 02 19:35:39 ERROR (1.595s)
```

mrshenli commented on torch/futures/__init__.py:
```diff
@@ -81,6 +81,39 @@ def then(self, callback):  # type: (Callable[[Future[T]], S]) -> Future[S]
+        added callbacks will be executed once the future completes.
+        If the future has already completed, the callback will be executed immediately.
+
+        Append the given callback function to this ``Future``, which will be run
```
There are some duplication between the two paragraphs. Is this intentional?

bdhirsh replied:
fixed, thanks :/

mrshenli commented on torch/futures/__init__.py:
```diff
@@ -81,6 +81,39 @@ def then(self, callback):  # type: (Callable[[Future[T]], S]) -> Future[S]
+        Append the given callback function to this ``Future``, which will be run
+        when the ``Future`` is completed.  Multiple callbacks can be added to
+        the same ``Future``, and will be invoked in the same order as they were
+        added. The callback must take one argument, which is the reference to
```
Shall we add a test where the callback function does not take any argument and verifies that the error message is clear?

bdhirsh replied:
Done. The callback logs exceptions to stdout, and I had trouble redirecting the content of stdout that was emitted from C++ in the python test so I could assert it's contents- right now the test just asserts that no exception was thrown. Here's an example though (showing that the exception is logged to stdout, but no exception is raised in the main thread)

![image](https://user-images.githubusercontent.com/16311747/95229334-730fba00-07ce-11eb-8b82-f7d099749a31.png)

mrshenli replied:
Can the following code solve this problem?

https://github.com/pytorch/pytorch/blob/275bb5e80139efc6f779b09d342b39eacccd97a3/torch/csrc/jit/python/pybind_utils.h#L127-L142

Looks like it works for existing tests:

https://github.com/pytorch/pytorch/blob/275bb5e80139efc6f779b09d342b39eacccd97a3/test/test_futures.py#L111-L116

bdhirsh replied:
I think the difference there is that when a callback registered with `then` encounters an exception, that exception gets bubbled up to the thread calling `then_fut.wait()` (making it easy to assert in the test).

Meanwhile the docs for `add_done_callback` assert that any exceptions encountered during the execution of the callback should be logged and ignored (https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Future.add_done_callback)- so we don't want to assert that we encountered an exception, we'd want to assert some string getting sent to stdout. I tried out this: https://stackoverflow.com/questions/4219717/how-to-assert-output-with-nosetest-unittest-in-python, but from some testing it only looks like it doesn't pick up anything that I log from C++ code (which is where I'm logging the exception).

bdhirsh replied:
I think the difference there is that when a callback registered with `then` encounters an exception, that exception gets bubbled up to the thread calling `then_fut.wait()` (making it easy to assert in the test).

Meanwhile the docs for `add_done_callback` assert that any exceptions encountered during the execution of the callback should be logged and ignored (https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Future.add_done_callback)- so we don't want to assert that we encountered an exception, we'd want to assert some string getting sent to stdout. I tried out this: https://stackoverflow.com/questions/4219717/how-to-assert-output-with-nosetest-unittest-in-python, but from some testing it only looks like it doesn't pick up anything that I log from C++ code (which is where I'm logging the exception).

rohan-varma commented on torch/testing/_internal/distributed/rpc/jit/rpc_test.py:
```diff
@@ -1038,6 +1038,32 @@ def callback(fut):
+        future_then = future.then(lambda _: True)
+
+
+        self.assertEqual(callback_called, None)
```
The callbacks should run when the `markCompleted` method is called on future, usually by the thread in the thread pool that processes the RPC response. Given this, isn't it theoretically possible the callback would've ran here and this assert could be flaky?

bdhirsh replied:
yep you're right, thanks for pointing that out. I think when I was writing corresponding tests between `test_futures.py` and `rpc_test.py` I missed the fact that "create the future" and "mark the future as complete" aren't split into two deterministic, separate steps. I'm just going to delete this assert, since I don't think that it adds much to the test (I could also move it to before the call to `add_done_callback` but I don't think that really adds anything).

bdhirsh committed:
Update on "fix #45552 - adding add_done_callback(fn) to torch.futures.Future"


Differential Revision: [D24055353](https://our.internmc.facebook.com/intern/diff/D24055353)

[ghstack-poisoned]

pritamdamania87 commented:
> Hey @pritamdamania87 this is requested by @jjlilley a few months ago in #37908, and this is also closer to the Python Future API. If we are not confident whether this will be useful, I guess we can mark this as a private API for now?

In general my feeling is that we should add a new API only if the existing one has gaps that cannot be filled by enhancing the existing API. Adding two similar APIs without a clear distinction between the two usually creates quite a bit of confusion. I guess we can keep this private for now since we've already spend quite a bit of effort here.

bdhirsh committed:
Update on "fix #45552 - adding add_done_callback(fn) to torch.futures.Future"


Differential Revision: [D24055353](https://our.internmc.facebook.com/intern/diff/D24055353)

[ghstack-poisoned]

bdhirsh committed:
Update on "fix #45552 - adding add_done_callback(fn) to torch.futures.Future"


Differential Revision: [D24055353](https://our.internmc.facebook.com/intern/diff/D24055353)

[ghstack-poisoned]

bdhirsh commented on torch/futures/__init__.py:
```diff
@@ -81,6 +81,39 @@ def then(self, callback):  # type: (Callable[[Future[T]], S]) -> Future[S]
+        this ``Future``. The callback function can use the ``Future.wait()`` API
+        to get the value.
+
+        We recommend that you use the ``then`` API as it provides a way to synchronize
```
Added some clarifications to the doc after talking with Greg. I don't really see a use case where `add_done_callback` gives you someone that `then` does not, so I added recommendations to the doc. I also clarified what actually happens when you interleave the two calls, since that's probably behaviors that users would want to know about (even though they probably shouldn't do that).

bdhirsh committed:
Update on "fix #45552 - adding add_done_callback(fn) to torch.futures.Future"


Differential Revision: [D24055353](https://our.internmc.facebook.com/intern/diff/D24055353)

[ghstack-poisoned]

mrshenli commented on torch/csrc/jit/python/pybind_utils.h:
```diff
@@ -144,6 +144,36 @@ struct VISIBILITY_HIDDEN PythonFutureWrapper
+              PyErr_Clear();
+            }
+            // Log and ignore exceptions raised through the callback
+            std::cout << "Got the following error when running the callback: "
```
usually, we don't do `cout`. If you would like to log, can we use sth like this:

https://github.com/pytorch/pytorch/blob/8cd3857bc7c57bd412941812f40fdb3a9dc6dd45/torch/csrc/distributed/rpc/tensorpipe_agent.cpp#L250

mrshenli commented on torch/futures/__init__.py:
```diff
@@ -81,6 +81,40 @@ def then(self, callback):  # type: (Callable[[Future[T]], S]) -> Future[S]
         return cast(Future[S], super().then(callback))
 
+    # Have to use string annotations because  PEP-0563 is not available in 3.6
+    def add_done_callback(self, callback):  # type: (Callable[[Future[T]], None]) -> None
```
As @pritamdamania87 suggested, let's make this API private by adding a "_" prefix to it.

bdhirsh replied:
done

mrshenli reviewed (approved):
Code looks good to me. But there are many irrelevant errors in tests. Could you please rerun tests?

bdhirsh committed:
Update on "fix #45552 - adding add_done_callback(fn) to torch.futures.Future"


Differential Revision: [D24055353](https://our.internmc.facebook.com/intern/diff/D24055353)

[ghstack-poisoned]

bdhirsh committed:
Update on "fix #45552 - adding add_done_callback(fn) to torch.futures.Future"


Differential Revision: [D24055353](https://our.internmc.facebook.com/intern/diff/D24055353)

[ghstack-poisoned]

bdhirsh committed:
Update on "fix #45552 - adding add_done_callback(fn) to torch.futures.Future"


Differential Revision: [D24055353](https://our.internmc.facebook.com/intern/diff/D24055353)

[ghstack-poisoned]

bdhirsh committed:
Update on "fix #45552 - adding add_done_callback(fn) to torch.futures.Future"


Differential Revision: [D24055353](https://our.internmc.facebook.com/intern/diff/D24055353)

[ghstack-poisoned]

bdhirsh committed:
Update on "fix #45552 - adding add_done_callback(fn) to torch.futures.Future"


Differential Revision: [D24055353](https://our.internmc.facebook.com/intern/diff/D24055353)

[ghstack-poisoned]

bdhirsh committed:
Update on "fix #45552 - adding add_done_callback(fn) to torch.futures.Future"


Differential Revision: [D24055353](https://our.internmc.facebook.com/intern/diff/D24055353)

[ghstack-poisoned]

rohan-varma commented on torch/csrc/jit/python/pybind_utils.h:
```diff
@@ -144,6 +144,36 @@ struct VISIBILITY_HIDDEN PythonFutureWrapper
+              // exception in the response message.
+              PyErr_Clear();
+            }
+            // Log and ignore exceptions raised through the callback
```
I might have missed some context here, but what's the reasoning behind `add_done_callback` swallowing exceptions but `then()` throwing the error? It seems like we should have consistency in the APIs here?

bdhirsh replied:
I just looked through some of the old comments in the PR, and I think the reason was because of the python documentation here: https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Future.add_done_callback. I wrote `add_done_callback` to mimic the same behavior that the python library has, which is to log but swallow exceptions. So I guess the question is whether it's worse to (a) have inconsistencies between our `add_done_callback` and python's , or (b) have inconsistencies between our `add_done_callback` and our `then`.

rohan-varma replied:
I actually did some more digging into this and I think what we have currently is fine. If you're interested I added an explanation to https://github.com/pytorch/pytorch/issues/52132
